extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -114,6 +131,43 @@ public void initialize() {
         }, 0, 1, TimeUnit.MINUTES);
     }
 
+    protected GatewayConfig getUsernameAndKey(GatewayConfig configuration) {
+        try {
+            DtlsConnectorConfig.Builder builder = new DtlsConnectorConfig.Builder(new InetSocketAddress(0));
+            builder.setPskStore(new StaticPskStore("Client_identity", configuration.code.getBytes()));
+
+            DTLSConnector dtlsConnector = new DTLSConnector(builder.build());
+            CoapEndpoint authEndpoint = new CoapEndpoint(dtlsConnector, NetworkConfig.getStandard());
+            String authUrl = "coaps://" + configuration.host + ":" + configuration.port + "/15011/9063";
+
+            CoapClient deviceClient = new CoapClient(new URI(authUrl));
+            deviceClient.setEndpoint(authEndpoint);
+
+            JsonObject json = new JsonObject();
+            String user = UUID.randomUUID().toString();
+            json.addProperty(CLIENT_IDENTITY_PROPOSED, user);
+
+            CoapResponse r = deviceClient.post(json.toString(), 0);
```

Comment:
How long could that take? Especially if the gateway is offline? Remember that initialize() has to return asap.

Reply:
I have already seen that on a working gateway initialize takes more than 5 sec - so I guess the entire initialization should be done async